src/compiler.ts

```diff
@@ -0,0 +1,160 @@
+import { filter, keys, clone, forEach, flow, negate, eq, includes, gte, gt, lte, lt, get, map, some, flatten, isString, isRegExp } from 'lodash/fp';
+import hi from './hidash';
+import { Query } from './types';
+
+type hasObjectQuery<T> = T extends Query ? true : false;
+type _compilePredicatesReturnType<T> = T extends true ? Array<(t: any) => boolean> : typeof clone<boolean>
+
+export default class QueryCompiler {
+  compile(query: Query) {
+    const filters = this._compilePredicates(query);
+    let filtered;
+    if (filters.length === 1) {
+      filtered = filter(filters[0]);
+    } else {
+      filtered = filters.map(function (v: any) {
+        return filter(v);
+      });
+      filtered = flow(filtered);
+    }
+    return filtered;
+  }
+
+  _compilePredicates<Q extends Query>(query: Q): _compilePredicatesReturnType<hasObjectQuery<Q>> {
+    let self = this;
+    let _keys = keys(query);
+    let filters: Array<(t: Query) => boolean> = [];
+    // if empty query, operation = clone collection
+    if (!_keys.length) {
+      return clone as _compilePredicatesReturnType<hasObjectQuery<Q>>;
```

Comment:
Is there another solution to this? Complains about is not `assignable` to the return type.